app/src/main/java/org/p2p/wallet/swap/SwapModule.kt

```diff
@@ -134,5 +148,61 @@ object SwapModule : InjectionModule {
         singleOf(::JupiterSwapTokensInMemoryRepository) bind JupiterSwapTokensLocalRepository::class
 
         factoryOf(::JupiterSwapInteractor)
+
+        factoryOf(::SwapStateRoutesRefresher)
+        factoryOf(::SwapWidgetMapper)
+        factoryOf(::SwapButtonMapper)
+
+        factory { (token: Token.Active?) ->
+            if (token == null) {
+                CommonSwapTokenSelector(
+                    get(),
+                    get(),
+                    get(),
+                )
+            } else {
+                PreinstallTokenASelector(
+                    get(),
+                    get(),
+                    get(),
+                    token,
+                )
+            }
+        }
+        factory { (token: Token.Active?) ->
+            SwapStateInitialLoadingHandler(
+                get(),
+                get(parameters = { parametersOf(token) })
+            )
+        }
+        factoryOf(::SwapStateLoadingRoutesHandler)
+        factoryOf(::SwapStateLoadingTransactionHandler)
+        factoryOf(::SwapStateSwapLoadedHandler)
+        factoryOf(::SwapStateTokenAZeroHandler)
+
+        factory { (token: Token.Active?) ->
+            setOf(
+                get<SwapStateInitialLoadingHandler>(parameters = { parametersOf(token) }),
+                get<SwapStateLoadingRoutesHandler>(),
+                get<SwapStateLoadingTransactionHandler>(),
+                get<SwapStateSwapLoadedHandler>(),
+                get<SwapStateTokenAZeroHandler>(),
+            )
+        }
+
+        factory { (token: Token.Active?, stateManagerHolderKey: String) ->
+            val stateManager = swapStateManagerHolder.getOrPut(stateManagerHolderKey) {
+                SwapStateManager(
+                    dispatchers = get(),
+                    handlers = get(parameters = { parametersOf(token) }),
+                )
+            }
+            JupiterSwapPresenter(
+                widgetMapper = get(),
+                buttonMapper = get(),
+                stateManager = stateManager
+            )
+        } bind JupiterSwapContract.Presenter::class
     }
+    val swapStateManagerHolder = mutableMapOf<String, SwapStateManager>()
```

Comment:
put it on the top please

Reply:
and also extract holder into a class and inject it everywhere it needs to be 
so we dont need to work with DI layer in presenter